src/Deinterleave.cpp

```diff
@@ -198,202 +384,345 @@ class Deinterleaver : public IRGraphMutator {
+                return Variable::make(t, op->name + ".lanes_1_of_3", op->image,
+                                      op->param, op->reduction_domain);
+            } else if (starting_lane == 2 && lane_stride == 3) {
+                return Variable::make(t, op->name + ".lanes_2_of_3", op->image,
+                                      op->param, op->reduction_domain);
             }
         }
+
+        // Uh-oh, we don't know how to deinterleave this vector expression.
+        // Make llvm do it.
+        return Shuffle::make({op}, lanes.indices);
     }
 
     Expr visit(const Cast *op) override {
         if (op->type.is_scalar()) {
             return op;
         } else {
-            Type t = op->type.with_lanes(new_lanes);
+            Type t = op->type.with_lanes(lanes.size());
             return Cast::make(t, mutate(op->value));
         }
     }
 
+    // FIXME: this is just broken.
     Expr visit(const Reinterpret *op) override {
+        // Don't mutate scalars
         if (op->type.is_scalar()) {
             return op;
         } else {
-            Type t = op->type.with_lanes(new_lanes);
+            // FIXME: Implement me.
+            Type t = op->type.with_lanes(lanes.size());
             return Reinterpret::make(t, mutate(op->value));
         }
     }
 
     Expr visit(const Call *op) override {
-        Type t = op->type.with_lanes(new_lanes);
+        Type t = op->type.with_lanes(lanes.size());
 
         // Don't mutate scalars
         if (op->type.is_scalar()) {
             return op;
         } else {
-
             // Vector calls are always parallel across the lanes, so we
             // can just deinterleave the args.
 
             // Beware of intrinsics for which this is not true!
             auto args = mutate(op->args);
-            return Call::make(t, op->name, args, op->call_type,
-                              op->func, op->value_index, op->image, op->param);
+            return Call::make(t, op->name, args, op->call_type, op->func,
+                              op->value_index, op->image, op->param);
         }
     }
 
     Expr visit(const Shuffle *op) override {
-        if (op->is_interleave()) {
-            // Special case where we can discard some of the vector arguments entirely.
-            internal_assert(starting_lane >= 0 && starting_lane < lane_stride);
-            if ((int)op->vectors.size() == lane_stride) {
-                return op->vectors[starting_lane];
-            } else if ((int)op->vectors.size() % lane_stride == 0) {
-                // Pick up every lane-stride vector.
-                std::vector<Expr> new_vectors(op->vectors.size() / lane_stride);
-                for (size_t i = 0; i < new_vectors.size(); i++) {
-                    new_vectors[i] = op->vectors[i * lane_stride + starting_lane];
-                }
-                return Shuffle::make_interleave(new_vectors);
-            }
+        // FIXME: are `op->vectors` guaranteed to not have duplicates?
```

Comment:
There's no such guarantee

Reply:
... the question being, should this code here be trying to deduplicate the `op->vectors`?

Reply:
I don't think so, no.